### ISSUE TITLE
NoWhitespaceBefore: Add support for closing brace of ARRAY_INIT separate from general RCURLY

### ISSUE THREAD

**Issue by eric-milles** [checkstyle/checkstyle#3326]
$ cat ArrayInit.java
```java
class ArrayInit
{
    String one = "value1", two = "value2";

    // no violations -- want single-line initializer to be compact
    String[] compact = {one};

    // 2 VIOLATIONS -- want single-line initializer to be compact
    String[] wantCompacted = { one };

    // no violations -- want multi-line initializer to be well-formed
    String[] justFine = {
        one,
        two,
    };
}
```

$ cat config.xml
```xml
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="NoWhitespaceAfter">
            <message key="ws.followed" value="Whitespace after ''{0}'' should be removed" />
            <property name="tokens" value="ARRAY_INIT" />
            <property name="allowLineBreaks" value="true" />
        </module>
        <!--module name="NoWhitespaceBefore">
            <message key="ws.preceded" value="Whitespace before ''{0}'' should be removed" />
            <property name="tokens" value="something that represents the RCURLY of an ARRAY_INIT" />
            <property name="allowLineBreaks" value="true" />
        </module-->
        <!-- This check is the complement to NoWhitespaceAfter ARRAY_INIT; the closing brace
             is just a plain RCURLY in the AST.  Goal is to produce tight array initializers
             (i.e. new int[] {1, 2}); but something will need to change when Lambdas show up. -->
        <module name="RegexpSinglelineJava">
            <property name="message" value="Whitespace before ''}'' should be removed" />
            <property name="format" value="(?:[^&quot;\s]\s+)\}" />
            <property name="ignoreComments" value="true" />
        </module>
    </module>
</module>
```

$ java -jar checkstyle-7.0-all.jar -c config.xml ArrayInit.java
Starting audit...
[ERROR] C:\Users\U0076777\ArrayInit.java:9: Whitespace before '}' should be removed [RegexpSinglelineJava]
[ERROR] C:\Users\U0076777\ArrayInit.java:9:31: Whitespace after '{' should be removed [NoWhitespaceAfter]
Audit done.
Checkstyle ends with 2 errors.

---------------

I have been missing the complement to NoWhitespaceAfter ARRAY_INIT for some time now.  NoWhitespaceBefore does not have a token type that matches the closing brace of an array initializer.  The desire is to produce compact array initializers when they are formatted to fit on one line, like `String[] compact = {"value"}` or `Arrays.asList(new String[] {"value"});` or `@SuppressWarnings({"rawtypes", "unchecked"})`.

Until now, I have been using a regex check to detect this condition (as seen in the config.xml above).  However, it is not very complete because we had been getting a lot of false positive matches within string literals.  Now that we have migrated to Java 8, the regex can't really be used anymore due to our choice of lambda syntax.

**Comment by romani** [comment]
```
java -jar checkstyle-6.17-all.jar -t ArrayInit.java 
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|--LITERAL_CLASS -> class [1:0]
|--IDENT -> ArrayInit [1:6]
`--OBJBLOCK -> OBJBLOCK [2:0]
    |--LCURLY -> { [2:0]
    |--VARIABLE_DEF -> VARIABLE_DEF [3:4]
    |   |--MODIFIERS -> MODIFIERS [3:4]
    |   |--TYPE -> TYPE [3:4]
    |   |   `--IDENT -> String [3:4]
    |   |--IDENT -> one [3:11]
    |   |--ASSIGN -> = [3:15]
    |   |   `--EXPR -> EXPR [3:17]
    |   |       `--STRING_LITERAL -> "value1" [3:17]
    |   `--SEMI -> ; [3:41]
    |--COMMA -> , [3:25]
    |--VARIABLE_DEF -> VARIABLE_DEF [3:4]
    |   |--MODIFIERS -> MODIFIERS [3:4]
    |   |--TYPE -> TYPE [3:4]
    |   |   `--IDENT -> String [3:4]
    |   |--IDENT -> two [3:27]
    |   `--ASSIGN -> = [3:31]
    |       `--EXPR -> EXPR [3:33]
    |           `--STRING_LITERAL -> "value2" [3:33]
    |--VARIABLE_DEF -> VARIABLE_DEF [6:10]
    |   |--MODIFIERS -> MODIFIERS [6:10]
    |   |--TYPE -> TYPE [6:10]
    |   |   `--ARRAY_DECLARATOR -> [ [6:10]
    |   |       |--IDENT -> String [6:4]
    |   |       `--RBRACK -> ] [6:11]
    |   |--IDENT -> compact [6:13]
    |   |--ASSIGN -> = [6:21]
    |   |   `--ARRAY_INIT -> { [6:23]
    |   |       |--EXPR -> EXPR [6:24]
    |   |       |   `--IDENT -> one [6:24]
    |   |       `--RCURLY -> } [6:27]
    |   `--SEMI -> ; [6:28]
    |--VARIABLE_DEF -> VARIABLE_DEF [9:10]
    |   |--MODIFIERS -> MODIFIERS [9:10]
    |   |--TYPE -> TYPE [9:10]
    |   |   `--ARRAY_DECLARATOR -> [ [9:10]
    |   |       |--IDENT -> String [9:4]
    |   |       `--RBRACK -> ] [9:11]
    |   |--IDENT -> wantCompacted [9:13]
    |   |--ASSIGN -> = [9:27]
    |   |   `--ARRAY_INIT -> { [9:29]
    |   |       |--EXPR -> EXPR [9:31]
    |   |       |   `--IDENT -> one [9:31]
    |   |       `--RCURLY -> } [9:35]
    |   `--SEMI -> ; [9:36]
    |--VARIABLE_DEF -> VARIABLE_DEF [12:10]
    |   |--MODIFIERS -> MODIFIERS [12:10]
    |   |--TYPE -> TYPE [12:10]
    |   |   `--ARRAY_DECLARATOR -> [ [12:10]
    |   |       |--IDENT -> String [12:4]
    |   |       `--RBRACK -> ] [12:11]
    |   |--IDENT -> justFine [12:13]
    |   |--ASSIGN -> = [12:22]
    |   |   `--ARRAY_INIT -> { [12:24]
    |   |       |--EXPR -> EXPR [13:8]
    |   |       |   `--IDENT -> one [13:8]
    |   |       |--COMMA -> , [13:11]
    |   |       |--EXPR -> EXPR [14:8]
    |   |       |   `--IDENT -> two [14:8]
    |   |       |--COMMA -> , [14:11]
    |   |       `--RCURLY -> } [15:4]
    |   `--SEMI -> ; [15:5]
    `--RCURLY -> } [16:0]

```


specific part:
```
`--ARRAY_INIT -> { [6:23]
    |--EXPR -> EXPR [6:24]
    |   `--IDENT -> one [6:24]
    `--RCURLY -> } [6:27]
```

NoWhitespaceBefore need to work with RCURLY token to demand this.
The similar case as in #3330 . We need special check for this .... NoWhitespaceIn might be good idea (Before, After, In).

If you need such Check please create new issue.

**Comment by eric-milles** [comment]
I understand from looking at the AST myself that there is no specific token to work with.  That is why I opened this issue.  What I am after is a specific token or something to set on the check to recognize the RCURLY that is part of ARRAY_INIT.

As I stated, NoWhitespaceAfter covers the opening brace.  I just need something in NoWhitespaceBefore to cover the closing brace.  These are the standard checks for this purpose.

**Comment by romani** [comment]
>What I am after is a specific token or something to set on the check to recognize the RCURLY that is part of ARRAY_INIT.

it is impossible to do. We have a tree and we can not change structure of it.
You requested smart treatment of RCURLY for ARRAY_INIT - so there will bunch of false positives.

if you know how to fix Check please do it and send PR, and testing report, .... .
I do not see good solution without new Check.

**Comment by eric-milles** [comment]
For my future understanding, if the AST does not have enough specificity for a given piece of code, there is little that can be done, correct?  That is, the AST is really only changing to add support for new syntax like Java 8 lambdas and so forth.

Is there the opportunity to specify a descendent token relationship in the config options?  Ex: ARRAY_INIT::RCURLY could be given to mean the RCURLY that is descendent from ARRAY_INIT.

What is the general policy for situations when a construct such as the array initializer needs a new check of its own, something like ArrayInitStyle (similar to AnnotationUseStyle) and some of the other checks overlap?  Typically are you just adding the new check and leaving any redundancies alone or deprecating them?  I am certainly happy to create a new check if there is high certainty of acceptance if I can meet your coding and testing guidelines.

Up to now, I have been trying to identify gaps in existing checks that have clear responsibility for specific code constructs.  WhetespaceBefore, WhatespaceAfter, and WhitespaceAround (as well as the No* variants) seem pretty clear to me.  Although I can understand the chance for complexity of implementation given the number of different situations they each handle.

**Comment by romani** [comment]
>  if the AST does not have enough specificity for a given piece of code, there is little that can be done, correct?

yes. We can not change existing structure of AST tree. It will result in damage in other Checks. There are hundred of them in Chekcstyle and even more thirdparty checks some where in web.
Java8 support demanded to change AST structure, so we take it for granted to retest/fix all Checks - that was unavoidable.

> Is there the opportunity to specify a descendent token relationship in the config options?

The only way to change it is to change the java grammar that is stored in checkstyle project. 

-----------------------

I will answer for all questions a bit later......

**Comment by eric-milles** [comment]
I created a check ArrayPad modeled after ParenPad.  It handles the spaces between `[]` and `{}` in array declarations, initializers, and dereferences.  I also have some handling for a pad option between the type and init list for the form `new Object[] {...}`, but I'm not sure if that belongs with this check or as part of [No]WhitespaceBefore ARRAY_INIT.

I've run it over a codebase of a few hundred thousand lines and it completed with no exceptions.  Can you give this a quick look and let me know if this is headed in the right direction.  I don't want to clone git and all that jazz unless it is worth it.

```java
package com.puppycrawl.tools.checkstyle.checks.whitespace;

import java.util.Locale;

import com.puppycrawl.tools.checkstyle.api.Check;
import com.puppycrawl.tools.checkstyle.api.DetailAST;
import com.puppycrawl.tools.checkstyle.api.TokenTypes;

/**
 * <p></p>
 *
 * <b>NOTE</b>: Whitespace after '{' may also be checked by: <pre>
 *  &lt;module name="NoWhitespaceAfter">
 *    &lt;property name="tokens" value="ARRAY_INIT" />
 *  &lt;/module></pre>
 *
 * <b>NOTE</b>: Whitespace between type or ident and '[' is checked by: <pre>
 *  &lt;module name="NoWhitespaceAfter">
 *    &lt;property name="tokens" value="ARRAY_DECLARATOR,INDEX_OP" />
 *  &lt;/module></pre>
 */
public class ArrayPad extends Check {
    protected PadOption option = PadOption.NOSPACE;

    public void setOption(String value) {
        option = PadOption.valueOf(value.trim().toUpperCase(Locale.ENGLISH));
    }

    @Override
    public int[] getDefaultTokens() {
        return new int[] {
            TokenTypes.ARRAY_DECLARATOR,
            TokenTypes.ARRAY_INIT,
            TokenTypes.INDEX_OP
        };
    }

    @Override
    public void visitToken(DetailAST opener) {
        DetailAST closer = opener.getLastChild();

        // limit checks to single line for simplicity
        if (opener.getLineNo() == closer.getLineNo()) {
            String line = getLine(opener);
            int next = opener.getColumnNo() + 1;
            int last = closer.getColumnNo() - 1;

            if (next <= last) { // not compacted
                if (option == PadOption.NOSPACE) {
                    if (isWhitespace(line, next)) {
                        log(opener.getLineNo(), next, "ws.followed", opener.getText()); //$NON-NLS-1$
                    }
                    if (isWhitespace(line, last) && next != last) { // don't log 2x for same char
                        log(closer.getLineNo(), last, "ws.preceded", closer.getText()); //$NON-NLS-1$
                    }
                }
                else if (option == PadOption.SPACE) {
                    if (isAllWhitespace(line, next, last)) {
                        log(opener.getLineNo(), next, "ws.followed", opener.getText()); //$NON-NLS-1$
                    }
                    else {
                        if (!isWhitespace(line, next)) {
                            log(opener.getLineNo(), next, "ws.notFollowed", opener.getText()); //$NON-NLS-1$
                        }
                        if (!isWhitespace(line, last)) {
                            log(closer.getLineNo(), last, "ws.notPreceded", closer.getText()); //$NON-NLS-1$
                        }
                    }
                }
            }
        }

        // check multi-line array initializers
        else if (opener.getType() == TokenTypes.ARRAY_INIT) {
            // check for an initializer expression that begins on the same line as the opening brace
            if (opener.getFirstChild() != null && opener.getFirstChild().getLineNo() == opener.getLineNo()) {
                String line = getLine(opener);
                int next = opener.getColumnNo() + 1;

                if (option == PadOption.NOSPACE && isWhitespace(line, next)) {
                    log(opener.getLineNo(), next, "ws.followed", opener.getText()); //$NON-NLS-1$
                }
                else if (option == PadOption.SPACE && !isWhitespace(line, next)) {
                    log(opener.getLineNo(), next, "ws.notFollowed", opener.getText()); //$NON-NLS-1$
                }
            }

            // check for an initializer expression that ends on the same line as the closing brace
            if (closer.getPreviousSibling() != null && closer.getPreviousSibling().getLineNo() == closer.getLineNo()) {
                String line = getLine(closer);
                int last = closer.getColumnNo() - 1;

                if (option == PadOption.NOSPACE && isWhitespace(line, last)) {
                    log(closer.getLineNo(), last, "ws.preceded", closer.getText()); //$NON-NLS-1$
                }
                else if (option == PadOption.SPACE && !isWhitespace(line, last)) {
                    log(closer.getLineNo(), last, "ws.notPreceded", closer.getText()); //$NON-NLS-1$
                }
            }
        }
    }

    protected final String getLine(DetailAST token) {
        return getLine(token.getLineNo() - 1);
    }

    /** Override to provide definition for "whitespace". */
    protected boolean isWhitespace(String line, int column) {
        return Character.isWhitespace(line.charAt(column));
    }

    protected final boolean isAllWhitespace(String line, int col1, int col2) {
        for (int i = col1; i <= col2; i += 1) {
            if (!isWhitespace(line, i)) {
                return false;
            }
        }
        return true;
    }
}
```

**Comment by romani** [comment]
> Is there the opportunity to specify a descendent token relationship in the config options?

yes it is possible by special option/property but in such a general Check any property will looks weird and related only to one token that might not be used at all.
New check is better.

> and some of the other checks overlap?

it is ok. General check are for general cases. Specific Check are always better to use.

> Typically are you just adding the new check and leaving any redundancies alone or deprecating them? 

Yes , we will keep redundacies alone.
no deprecation is required. It is up to user what to use, we could make a reference to other Check in documentations of general Checks.

>  I am certainly happy to create a new check if there is high certainty of acceptance if I can meet your coding and testing guidelines.

I can not guaranty , testing will show how good your proposal.
For now I like your idea in general, but I need more details on USAGE not an implementation.
Please move your new Check idea to sevntu project - https://github.com/sevntu-checkstyle/sevntu.checkstyle/ and describe it from usage perspective: config view + code ==> violations.
Please create new issue and lets do discussion from scratch about new idea.
